pkg/logs/input/docker/matcher.go

```diff
@@ -47,10 +40,35 @@ func (s *headerMatcher) Match(exists []byte, appender []byte, start int, end int
 // case [1|2 0 0 0 size1 size2 size3 10]
 func (s *headerMatcher) matchHeader(exists []byte, bs []byte) bool {
 	l := len(exists) + len(bs)
-	if l >= headerLength || l < headerPrefixLength {
+	if l < headerPrefixLength {
 		return false
 	}
-	h := append(exists, bs...)
-	return bytes.HasPrefix(h, headerStdoutPrefix) ||
-		bytes.HasPrefix(h, headerStderrPrefix)
+
+	for i := 0; i < 4; i++ {
+		// possible start of header offset
+		idx := l - (headerPrefixLength + i)
+		if idx < 0 {
+			// less than 4 + i bytes
+			continue
+		}
+		if s.getByte(exists, bs, idx) == 1 || s.getByte(exists, bs, idx) == 2 {
+			if s.getByte(exists, bs, idx+1) == 0 &&
+				s.getByte(exists, bs, idx+2) == 0 &&
+				s.getByte(exists, bs, idx+3) == 0 {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (s *headerMatcher) getByte(exists []byte, bs []byte, i int) byte {
+	l := len(exists) + len(bs)
+	if i < l {
+		if i < len(exists) {
+			return exists[i]
+		}
+		return bs[i-len(exists)]
+	}
+	return 0xFF
```

Comment:
I've switch `getByte` to `checkByte` and wrap the equality assertion inside the function. IMHO it keeps the code simple while avoiding the case where it could return 0xFF for a non-existing byte.